### PR TITLE
 rose macro, app-upgrade: handle opt confs.

### LINKED ISSUE #1334: rose app-upgrade, rose macro: handle optional configurations?
We should see if it is possible to support optional configurations in macros.

### PR DESCRIPTION
**Pull request by benfitzpatrick**
We can now handle validation and macro transformation of
optional configurations in rose macro and rose app-upgrade.

The commands now support `-O` and `--opt-conf-key` options.
The old `output_dir` option has been changed to use `-o` rather
than `-O`.

This closes #1334.

Validation is easy.

Transforms are hard. It is difficult to know what to apply
where. It currently works like this:
 * add setting in macro => add settings to the main configuration, unless
 their parent section only exists in an optional configuration
 * modify setting in macro => modify setting only in the last applied
 configuration containing that setting (so e.g. in the optional
 configuration rather than the main one, which would get overwritten).
 * remove setting in macro => remove setting from all configurations

In the case of modification, this may leave the main configuration
in an 'un-upgraded' state if the setting was present in the optional
configuration as well. However, it's difficult to get these things
right - arguably, the best thing for the user to do is to upgrade with
optional configurations, revert the main configuration, and then just
upgrade the main configuration.

@matthewrmshin, please review.
@dpmatthews, I know you have suggestions for the desired mechanics.

### CODE REVIEW

File: lib/python/rose/config.py [hunk 1]
```diff
@@ -278,6 +279,116 @@ def unset(self, keys=None):
...
+
+
+class ConfigNodeDiff(object):
+
+    """Represent differences between two ConfigNode instances."""
+
+    def __init__(self):
+        self._data = {"added": {}, "removed": {}, "modified": {}}
+
+    def set_from_configs(self, first_config_node, second_config_node):
```
Comment (matthewrmshin) [inline lib/python/rose/config.py]: How about `config_node_1`, `config_node_2`? Shorter and easier for the eyes.

**Comment by matthewrmshin** [comment]
Test battery OK. We probably need a quick truth table for the behaviour.

**Comment by benfitzpatrick** [comment]
99a60e6 adds a truth table and fixes the naming comment.

**Comment by matthewrmshin** [comment]
@benfitzpatrick as discussed this morning, please evolve this towards the @dpmatthews 's idea.

**Comment by benfitzpatrick** [comment]
Now updated with new mechanisms.

**Comment by benfitzpatrick** [comment]
Reference guide entry added.

**Comment by matthewrmshin** [comment]
Test battery OK in my environment.

File: lib/python/rose/config.py [hunk 2]
```diff
@@ -278,6 +279,138 @@ def unset(self, keys=None):
...
+    def __init__(self):
+        self._data = {"added": {}, "removed": {}, "modified": {}}
+
+    def set_from_configs(self, config_node_1, config_node_2):
+        """Create diff data from two ConfigNode instances."""
+        first_settings = {}
+        second_settings = {}
+        for config_node, settings in [(config_node_1, first_settings),
+                                      (config_node_2, second_settings)]:
+            for keys, node in config_node.walk():
```
Comment (matthewrmshin) [inline lib/python/rose/config.py]: Should probably match names here:

```python
[(config_node_1, settings_1), (config_node_2, settings_2)]
```

File: lib/python/rose/config.py [hunk 3]
```diff
@@ -278,6 +279,138 @@ def unset(self, keys=None):
...
+        for keys, info in self.get_removed():
+            # Need to add as user-ignored.
+            value, state, comments = info
+            node.set(keys, value=value, state=node.STATE_USER_IGNORED,
+                     comments=comments)
+        return node
+
+    def set_added_setting(self, keys, data):
+        """Set a setting to be "added"."""
+        self._data["added"][keys] = data
```
Comment (matthewrmshin) [inline lib/python/rose/config.py]: Use constants for `"added"`, `"modified"`, `"removed"` etc.

**Comment by kaday** [comment]
Passes test-battery in my environment. Looks okay to me, but agree with @matthewrmshin suggestions.

**Pull request merged**